### ISSUE TITLE
Memory leak for spectral analysis

### ISSUE THREAD

**Issue by ctrichard** [gammapy/gammapy#1213]
Dear All,

As mentioned during last Gammapy call, I have a memory leak during several analyses of CTA DC1.

For around 40 sources to analyse I end with jobs of around 1.5 - 2 Go in memory...

Cheers,
Cyril

**Comment by cdeil** [comment]
This is a known and major issue in Gammapy that is known at least to @joleroi @adonath @registerrier  and me and that we have discussed before. There are some analyses in Gammapy that will load more and more data (EVENTS and IRFs) into memory and never unload it. This has worked for us so far, but for deep CTA observations it leads to many GB (or even 10s of GB for survey analyses) and the Python process usually starts swapping to disk and becomes really slow and effectively runs forever and clogs the user machine. Very bad.

The problem is that it requires quite large code changes and cleanup to resolve and we haven't spent the time yet to find and agree the good new solution how to do data loading and unloading from analysis pipelines. But we have to resolve this now and start work on something better!

The core of the issue is that we have high-level analysis "pipeline" classes that take an [gammapy.data.ObservationList](http://docs.gammapy.org/en/latest/api/gammapy.data.ObservationList.html) as input, usually obtained from [DataStore.obs_list](http://docs.gammapy.org/en/latest/api/gammapy.data.DataStore.html#gammapy.data.DataStore.obs_list).

I think this is the list of high-level analysis classes are affected (list might be incomplete):
*  [SpectrumAnalysisIACT](http://docs.gammapy.org/en/latest/api/gammapy.scripts.SpectrumAnalysisIACT.html)
 * [SpectrumExtraction](http://docs.gammapy.org/en/latest/api/gammapy.spectrum.SpectrumExtraction.html)
*  [IACTBasicImageEstimator.run](http://docs.gammapy.org/en/latest/api/gammapy.image.IACTBasicImageEstimator.html#gammapy.image.IACTBasicImageEstimator.run)

What happens is that this is a list of [gammapy.data.DataStoreObservation](http://docs.gammapy.org/en/latest/api/gammapy.data.DataStoreObservation.html) classes, where a `DataStoreObservation` is a reference to a `DataStore` and an `obs_id`, so a proxy object that can load EVENTS and IRFs as-needed, i.e. when someone accesses e.g. the `.events`, `.aeff` etc attributes they are loaded and cached (these are [astropy.utils.decorators.lazyproperty](http://docs.astropy.org/en/stable/api/astropy.utils.decorators.lazyproperty.html)). The idea was to avoid inefficiencies of multiple FITS file open and re-read with the convenience of just accessing what you need. But there is no concept of unload, and the `ObservationList` and of caching attributes on `DataStoreObservation` conspire so that any loaded EVENT or IRF is never released by the Python garbage collection (the list keeps a reference to each obs, and the obs caches it's data members).

@adonath @joleroi @registerrier all - What would you propose? Suggestions for short-term fixes or long-term solution welcome!

One solution I like is to rewrite the high-level analysis classes to look like this:
https://github.com/gammasky/cta-dc1-gps-analysis/blob/d91345301665cef41986f786c127578b03708339/make_cube.py#L64
```
    data_store = get_datastore()
    cube = SkyCube.empty_like(image_ref, energies=e_bin['e_bounds'])

    for obs_id in tqdm(obs_ids):
        obs = data_store.obs(obs_id)

        cutout_cube, cutout_slices = make_cutout(cube, obs.pointing_radec, offset_max)

        obs_cube = make_counts_cube(
            event_list=obs.events,
            ref_cube=cutout_cube,
            pointing=obs.pointing_radec,
            offset_max=offset_max,
        )

        cube.data[cutout_slices] += obs_cube.data
```
If we do this we can delete the `ObservationList` class. This works and is simple, all data associated with a given `obs` is garbage collected at the end of the for loop. It does have the problem that it couples our high-level analysis code to the `DataStore`, which I agree is not desirable and an interim abstraction like `ObservationList` that could be created without any index files and DataStore would be better.

Another alternative would be that we change the `lazyproperty` for `obs.events` etc to `property`, i.e. remove the caching. I think this might resolve the issue, without any changes needed for the high-level analysis classes? And most (all?) of our analysis classes currently don't access the same thing multiple times, so the `lazyproperty` was a premature optimisation (or rather de-optimisation)? I'm not sure if this is true, or if some Python references to the objects and the memory issue will remain. One quick way to check would be to run an example via https://github.com/astrofrog/psrecord for 100 or 1000 CTA runs and see if the Python process memory keeps growing and growing or if it stays constant (for me at 300 MB on my Macbook).

So a quick and good fix might be possible here. But ideally someone would take on this "project" of analysis efficiency and data loading / unloading checks and improve the existing code, docs and tests we have. I'll leave this issue open until next week, and if no-one becomes active, will start working on this myself.

**Comment by adonath** [comment]
I ran into the same issue with the `IACTBasicImageEstimator` and my quick and dirty fix was just to manually delete the event list after it's not needed any more (see [here](https://github.com/gammapy/gammapy/blob/master/gammapy/image/basic.py#L311)).

Conceptionally I really like the concept of the `ObservationList` and I would prefer to keep this class (literally speaking the idea is that you go into the data store, you get your bag with observations according to your selection and preferences and you don't carry the whole store with you all the time...).

The problem right now is, that `ObservationList` is a `UserList` and when used as an iterator, all observations (even when they are not used anymore) are kept in memory. My favourite solution would be to implement a generator method for the `ObservationList` class, so that every item is only used once and is thrown away afterwards. This allows us to keep the nice coding pattern with loops over actual observations (and not observations ids), while solving the issue with the memory leakage.
I think @kosack uses generators in ctapipe as well for loops over single events.

**Comment by cdeil** [comment]
@adonath - A generator doesn't let you access a given obs_id directly if you want, and many people are confused that you can iterate over it only once and then it's empty. I think exposing it to users could lead to a lot of problems.

How about keeping the `ObservationList` for now and removing the caching on the obs properties? Do you see any problem with that?

And then in the future (after more thought and discussion and maybe even prototyping) we write new Observation and Observations classes. That would behave a lot like the current `DataStoreObservation` and `ObservationList` classes, i.e. high-level analysis code would still work with those abstractions, but they wouldn't be coupled to the datastore any more, creating them via the datastore would be just one option.

**Comment by cdeil** [comment]
One more thought, in case we don't want to remove the caching on obs properties: `ObservationList` could be replaced by a class `Observations`, that (for now) has a list of `obs_id` as data member as well as a data store, and just offset a `get(obs_id)` interface, but doesn't hold a reference to the obs itself.

I.e. basically it would just be a bundling of list of obs_id and datastore to avoid introducing those in high-level analysis interfaces, and it's future-proof because it only has a minimal access interface that in the future a better `Observations` class would also have (it would know the list of observations it contains and let you get an obervation). Then high-level analysis code could keep things loaded by storing it if they want, or in most cases they would remove their reference to the obs object after processing one obs and moving on to the next one. In that case we could keep using `lazyproperty` for the obs properties.

But is this a better way to proceed for now?
I'm leaning towards doing `lazyproperty` -> `property` and then the `del` hack in your `IACTImageEstimator` can also be removed.

Do we have any high-level analysis code that accesses obs properties like `obs.events` or `obs.aeff` more than once (or twice)?

**Comment by kosack** [comment]
I agree you should probably use generators (if it's a single function), or a class that implements the iterator interface (if it's a class).   In the latter, you just add `__iter__()` and  `__next__()` methods to your class that do the loading of a single run upon request.  You don't even need to give up random access (though I can't see quite where you need it), you just handle loading the data when the item is requested, and not loading it all at once. 

Generators are just a nice way to write something like an iterator class without writing a class, only using a single function that emulates a list, but loads items lazily. 

Conceptually it would look like the following for the generator method:

```py3
def get_obs_data_list(obs_ids):
     for run in runs:
         yield load_obs_data(obs_id)  # loads a single run into memory

# then the analysis loop:
allrundata = get_obs_data_list(obs_ids)

for obs_data in all_obs_data:
     process(obs_data)

# or even:
map(process, obsdata)
```



or even shorter if you use a generator comprehension
```py3
allrundata = ( load_obs_data(x) for x in obs_ids )
```

**Comment by kosack** [comment]
>access a given obs_id directly if you want, and many people are confused that you can iterate over it only once and then it's empty. I think exposing it to users could lead to a lot of problems

For this, I would recommend just implementing `__iter__()`, `__next__()`, and `__getitem__()` in your class, and internally make it do lazy data loading.  Then you get the benefit of a generator, as well as random access. You just have to make sure you don't keep the resulting single observation data in a variable somewhere in a class that uses it, otherwise the garbage collector will not delete it, and you'll have another memory leak.  E.g. always pass around the ObservationList, and never store observations locally.

**Comment by cdeil** [comment]
I talked to @adonath about this and the short-term plan is this:

- Add a Python script that shows the ever-increasing memory when running a spectral extraction in https://github.com/gammapy/gammapy-benchmarks
- Ideally, run it through https://github.com/astrofrog/psrecord and store a CVS file and / or plot that shows the current performance issue and commit that.
- Then change `lazyproperty` to `property` on the `DataStoreObservation` to fix the issue for now in Gammapy
- Re-run the high-level performance test and commit the new version (memory should be stable at 400 MB).

So this is the start of high-level performance benchmarks for Gammapy.

@adonath - If you want to do it quickly, access CTA DC-1 data. If you have a little time, generating test data would be nicer. I think it should be possible to just create 100 or 100 dummy run datasets by creating multiple copies e.g. of https://github.com/gammapy/gammapy-extra/tree/master/test_datasets/cta_1dc via `make.py test-data`. Or actually I think we don't need to duplicate the data files at all, we can just have an index file that points to the same single file 100 or 1000 times, and the way our data loading works is that each obs will re-load the data, because it doesn't know that it's always the same EVENTS and IRF data files.

**Comment by cdeil** [comment]
PR to change DataStoreObservation from lazy to non-lazy properties: #1226

This is just a small step that resolves the issue for now. I would suggest we keep this issue open as a reminder and to discuss further related changes to data loading. Changing the milestone here to v0.8.

**Comment by cdeil** [comment]
I'm not sure if we should close the issue or not. The current solution of not caching data on the observation objects basically resolves it. The only problem with that is that possibly data (i.e. EVENTS and IRFs) can be read multiple times from disk, depending on how analysis code is structured.

@registerrier - In your upcoming proposal for code that makes cubes / maps / stacked PSF and EDISP, would you read data once or multiple times? I guess AEFF is read multiple times, e.g. in the function that computes exposure, and then again in the PSF and EDISP computation? Or is this avoided somehow by storing AEFF or exposure in the higher-level analysis code that computes the stacked responses?

In any case, I'm not sure if we should add a caching mechanism to the observation or observationlist class, i.e. introduce complexity to reduce I/O. @registerrier - thoughts?

**Comment by registerrier** [comment]
The low level functions all take simple inputs (``EventList``, ``EnergyDispersionTable2D`` etc). This issue has to be tackled by higher level analysis scripts/classes.
A priori, map making and e.g. PSF kernel extraction can be independent, in which case one has to read IRFs multiple time.

**Comment by cdeil** [comment]
@registerrier - Close issue or keep open?

I want to release Gammapy v0.8, so moving milestone to v0.9 here.

**Comment by cdeil** [comment]
This was resolved in #1226 .  We will be developing new data classes in the coming months and when objects are created and deleted will be a big topic. But I don't see any remaining concrete action item here, so closing this issue now.
